dht-cache/src/cache.rs

```diff
@@ -0,0 +1,462 @@
+                            .await
+                            .ok()
+                            .map(|_| DomoEvent::PersistentData(elem))
+                    } else {
+                        None
+                    }
+                }
+            }
+        }
+    });
+
+    (cache, events)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::dht::test::*;
+    use std::{collections::HashSet, pin::pin};
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 3)]
+    async fn builder() {
+        let cfg = crate::Config {
+            shared_key: "d061545647652562b4648f52e8373b3a417fc0df56c332154460da1801b341e9"
+                .to_owned(),
+            ..Default::default()
+        };
+
+        let (_cache, _events) = Builder::from_config(cfg).make_channel().await.unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+    async fn syncronization() {
+        let [mut a, mut b, mut c] = make_peers(2).await;
+        let mut d = make_peer(2).await;
+
+        connect_peer(&mut a, &mut d).await;
+        connect_peer(&mut b, &mut d).await;
+        connect_peer(&mut c, &mut d).await;
+
+        let a_local_cache = LocalCache::new();
+        let b_local_cache = LocalCache::new();
+        let c_local_cache = LocalCache::new();
+        let d_local_cache = LocalCache::new();
+
+        let mut expected: HashSet<_> = (0..10)
+            .into_iter()
+            .map(|uuid| format!("uuid-{uuid}"))
+            .collect();
+
+        let (a_c, a_ev) = cache_channel(a_local_cache, a, 100);
+        let (b_c, b_ev) = cache_channel(b_local_cache, b, 100);
+        let (c_c, c_ev) = cache_channel(c_local_cache, c, 100);
+
+        let mut expected_peers = HashSet::new();
+        expected_peers.insert(a_c.peer_id.clone());
+        expected_peers.insert(b_c.peer_id.clone());
+        expected_peers.insert(c_c.peer_id.clone());
+
+        tokio::task::spawn(async move {
+            let mut a_ev = pin!(a_ev);
+            let mut b_ev = pin!(b_ev);
+            let mut c_ev = pin!(c_ev);
+            for uuid in 0..10 {
+                let _ = a_c
+                    .put(
+                        "Topic",
+                        &format!("uuid-{uuid}"),
+                        &serde_json::json!({"key": uuid}),
+                    )
+                    .await;
+            }
+
+            loop {
+                let (node, ev) = tokio::select! {
+                    v = a_ev.next() => ("a", v.unwrap()),
+                    v = b_ev.next() => ("b", v.unwrap()),
+                    v = c_ev.next() => ("c", v.unwrap()),
+                };
+
+                match ev {
+                    DomoEvent::PersistentData(data) => {
+                        log::debug!("{node}: Got data {data:?}");
+                    }
+                    _ => {
+                        log::debug!("{node}: Other {ev:?}");
+                    }
+                }
+            }
+        });
```

Comment:
This is not cancel safe, and this is the reason we should avoid using `select!` when possible. This is a possible approach using `futures-concurrency`:

```rust
            (
                a_ev.map(|v| ("a", v)),
                b_ev.map(|v| ("b", v)),
                c_ev.map(|v| ("c", v)),
            )
                .merge()
                .for_each(|(node, ev)| async move {
                    match ev {
                        DomoEvent::PersistentData(data) => {
                            log::debug!("{node}: Got data {data:?}");
                        }
                        _ => {
                            log::debug!("{node}: Other {ev:?}");
                        }
                    }
                })
                .await;
```

Reply:
You are right, the stream accesses RwLocks, it should be also documented